### ISSUE TITLE
Transactions not showing up in sentry when using manual instrumentation.

### ISSUE THREAD

**Issue by Allieway** [getsentry/sentry-php#1362]
### How do you use Sentry?

Sentry Saas (sentry.io)

### Version

newest

### Steps to Reproduce
I might be doing something wrong but my code looks correct, this is a lot of info, see next bolded text for TL;DR

I have a trace that starts with a transaction and has 13 spans within in, some span children of the transaction, and other span children of those children. 
I close spans in order, each span has the correct tags thanks to my array_merge before setting context tags. 
I set the description of the parent transaction to the url path I'm implementing manual instrumentation on, and each other spans description to it's index in the full trace.
But the description of the top level span is set to "Index 13" -- but so is the last child span "Index 13"
The tags in the main section of the event details are the same ones as the child "Index 13", as are all the other attributes, including "parentSpanId". 
The root parent has no tags when I click on it in event details, but it does have the same "parentSpanId" and other default filled fields. 
When I go to the "Trace View" page it doesn't show children, only the root span. That root span is "Index 13" and does not have a "parentSpanId" property on that page view, but the "spanId" property matches the last child span with the same description(and the root span that I believe might be working incorrectly). 
I var_dumped as far as the eye can see. Before and After I modified any context(transaction and span), before and after I modified the transaction, before and after I modified all spans, and before and after I called ->finish() on spans and transactions. 
After ->finish() is called on the transaction object, the var_dump of that object is the very last output on the page. It has the correct tags, properties, name, and ID. It has an endTimeStamp and sampled is true.
The only reference to that ID on the event details page is in the parentSpanId field. There are no properties, name, or tags that match my output of the root transaction object after->finish(); is called(or anytime that transaction is output) on that same page. 

**TL;DR**:
Something is happening that causes the root span(transaction?) to have the properties of the last child span on the transaction summary page. Even the "Trace Details" section of the event contain the last spans properties, including parentSpanId. 
Keep in mind, I touched PHP, Sentry, and APM in general all for the first time 2 weeks ago, so it's possible I'm hugely misunderstanding what's going on. Below is the transaction output with details scrubbed, I may contact sentry support as well? Honestly I'm not sure what I should do at this point. 

```
object(Sentry\Tracing\Transaction)
private 'name' => string 'Nowhere to be found on event details page'
protected 'spanId' => 
    object(Sentry\Tracing\SpanId)
      private 'value' => string 'only visible as parentSpanId property'
protected 'traceId' => 
    object(Sentry\Tracing\TraceId)
      private 'value' => string 'same ID I see for traceId on event details page'
  protected 'status' => 
    object(Sentry\Tracing\SpanStatus)
      private 'value' => string 'ok'
  protected 'parentSpanId' => null
  protected 'sampled' => boolean true
  protected 'tags' => 
    array (size=16)
    'a bunch of key' => 'value pairs that are not in sentry UI'
  'protected 'endTimestamp' => float 'unix time float filled in'
protected 'spanRecorder' => 
    object(Sentry\Tracing\SpanRecorder)
      private 'maxSpans' => int 1000
      private 'spans' => 
        array (size=13)
          0 => 
            &object(Sentry\Tracing\Transaction)
          1 => 
            object(Sentry\Tracing\Span)
              ...
          2 => 
            object(Sentry\Tracing\Span)
              ...
          3 => 
            object(Sentry\Tracing\Span)
              ...
          4 => 
            object(Sentry\Tracing\Span)
              ...
          5 => 
            object(Sentry\Tracing\Span)
              ...
          6 => 
            object(Sentry\Tracing\Span)
              ...
          7 => 
            object(Sentry\Tracing\Span)
              ...
          8 => 
            object(Sentry\Tracing\Span)
              ...
          9 => 
            object(Sentry\Tracing\Span)
              ...
          10 => 
            object(Sentry\Tracing\Span)
              ...
          11 => 
            object(Sentry\Tracing\Span)
              ...
          12 => 
            object(Sentry\Tracing\Span)
```

the 12th index in spanRecorder is what I call index 13(count from 1) and its details are displayed for the transaction and general trace details properties. 
I'll keep digging, but would appreciate assistance. 

### Expected Result
Have transaction properties be set in the root span of the overall trace and be in trace details section. 

### Actual Result
The last child span's properties are set as the root span of the overall trace and in the trace details section.

**Comment by Allieway** [comment]
I spent all day trying to troubleshoot. I have been comparing Sentry Go SDK and PHP SDK a ton. They do a lot of things differently. But I did make progress. 

[if ($span->getSpanId() !== $this->getSpanId() && null !== $span->getEndTimestamp()) {](https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/Tracing/Transaction.php#L96)

That needs to be changed to:
`if (null !== $span->getEndTimestamp() {)`

The event uses getSpans() to set the spans object while building the envelope. 
We run this:
`$event->setSpans($finishedSpans);`
To set the spans. But we don't add the transaction to the list of spans.

The only thing we're passing from the transaction object to the event is the name. 
I love that the Transaction Property randomly turns into a string when we go to the event. I understand there is some open telemetry stuff involved or something else that makes the transaction field in the payload be the name. But it really doesn't make logical sense.

Anyway, 
I dug in deep and ended up being here:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/Transport/HttpTransport.php#L116
I kept going and ended up here:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/Serializer/PayloadSerializer.php#L162

The data looked correct there. It had the transaction at the top. It had all required fields. 

But then it goes and creates a stream. and Idk what happens after that. 

After I fixed it so that the transaction is actually included in the payload... All the child spans are correctly placed underneath the transaction in sentry UI. The tags are set correctly, and all the properties. 

So horray It's fixed!!!
right?

Nope, the last child span is somehow above the transaction in the UI now. Which means all the root trace details are still being set to that last child's details, and it's not super functional.

**Comment by cleptric** [comment]
Hey

I would recommend separating different topics you have into different GitHub issues, so it's easier for us to give feedback.

I understand that SpanContext::getTags and Span::getTags do not contain the same data. You could open a PR that implements a fix or just a failing test so that we can take it from there.

About your second topic, I have a hard time fully grasping the issue. I would recommend you share some code on how you set up your instrumentation or share a repository with sample code that showcases what you did.

**Comment by Allieway** [comment]
@cleptric 

Literally the 1st sentence says the 1st issue is with setTags, not getTags. I also specified the exact lines that need changing. I really don't think anyone who knows this codebase would need a failing test to take if from where I left it(literally 1 line change, and what that change is). 

If you can't grasp the 2nd issue with the amount of detailed information I provided, then you won't be able to grasp 1000 lines of implementation code.

I agree this needs reproduction, but it does not need more information. 

I appreciate the response, but I'd rather this sit in untriaged status longer with no response than shifting blame to the issue creator with a needs information status and a response that displays a lack of reading comprehension across the board. 

If this was triaged correctly, it would have a status of in progress, unrouted, confirmed, or help wanted. With a team of web backend. And a type of breaking, bug, question, or support request. 

If I'm correct, this affects every single person using manual instrumentation and has been for a long time, and affects them in a big way. If I'm not, then the fact that it has that possibility alone should warrant a response by someone that at the very least spends a sliver of the amount of time I spent into digging into this issue, looking into it themselves. 

I'm genuinely sorry for being an asshole in general. I just really don't think the statuses you set and the place you left this issue are the correct ones, and that is hasn't actually been triaged.

**Comment by Allieway** [comment]
Separated(unbelievably low priority) issue is here:
https://github.com/getsentry/sentry-php/issues/1368

**Comment by Allieway** [comment]
I got it fixed. The adding of the transaction to the spans made it show up, but wasn't the correct direction to go. I ended up finding out that contexts.trace is supposed to contain the transaction details, and it is supposed to be excluded from spans. 

We set the contexts.trace property to the transaction properties here: 
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/Tracing/Transaction.php#L108

Then we call CaptureEvent:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/State/Hub.php#L147-L156

Capture event passes the scope received from these lines to captureEvent in the client.
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/State/Hub.php#L294-L305
I'm unsure how the stacks or layers work and I wasn't sure how to interact with $hub->stack to see what the object looks like. But it looks to grab the last item, and that item in my scenario was the last child span's scope.

Anyway, that scope is passed to client->prepareEvent() here:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/Client.php#L145-L147

Inside prepareEvent here is where the scope is used:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/Client.php#L262-L271

We are now in scope->applyToEvent() with the scope of the last child span. 
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/State/Scope.php#L323-L353
All of those lines only set event properties with scope properties if they aren't set yet, or they merge event and scope properties together, with preference for the event. 

But these lines are different:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/State/Scope.php#L355-L358
The first if statement sets the events contexts(trace specifically) to the results of span->getTraceContext();

This is what that looks like:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/Tracing/Span.php#L336-L368
Since the span we're calling that on is the last span, it sets all of the events trace contexts to the last span's properties. 

Then we get here:
https://github.com/getsentry/sentry-php/blob/c0037a571336722a8853068d56e2e6ecfdf1ee1f/src/State/Scope.php#L360-L362
It's safe again, merging arrays, and not overwriting anything that's already set on the event.

I don't see where context is set on the scope. I see setContext public function, but I don't see that called anywhere. 
Which leads me to believe that this:
`$event->setContext('trace', $this->span->getTraceContext());`
Was intended/should be this:
`$this->setContext('trace', $this->span->getTraceContext());`

Making that change fixed everything completely. I have a super clean trace event details page with all properties set correctly. 

With that said, you would think that span->getTraceContext() should get the transaction's properties, not any spans properties that it's called on. So maybe the setting of the event context with getTraceContext was intentional? And getTraceContext should give back the correct information? I'm not sure.

**Comment by Allieway** [comment]
Created a PR, But there seem to be a couple ways to implement the fix and I'm not sure what the best approach is, so I just did the easiest one.
